### PR TITLE
[ML] Fix potential problem for boosted tree training initialisation

### PR DESCRIPTION
**Pull request by tveasey**
Numeric errors mean that it's possible the sum curvature for a candidate split is identically zero while the gradient is epsilon. This can cause the node gain to appear infinite (when there is no weight regularisation) which in turns causes problems initialising the region we search for optimal hyperparameter values.

Since none of our loss functions should have curvature identically equal to zero, we can safely force the gradient to be zero if detect that the curvature is zero.

\cc @przemekwitek.

### CODE REVIEW

**Comment by tveasey** [comment]
No problem! Thanks for raising and helping investigate @przemekwitek.

File: lib/maths/CBoostedTreeImpl.cc [hunk 1]
```diff
@@ -110,13 +110,18 @@ CBoostedTreeImpl::CLeafNodeStatistics::CLeafNodeStatistics(std::size_t id,
         m_Gradients[i].resize(numberSplits);
         m_Curvatures[i].resize(numberSplits);
         for (std::size_t j = 0; j < numberSplits; ++j) {
-            m_Gradients[i][j] = parent.m_Gradients[i][j] - sibling.m_Gradients[i][j];
             m_Curvatures[i][j] = parent.m_Curvatures[i][j] - sibling.m_Curvatures[i][j];
+            m_Gradients[i][j] = m_Curvatures[i][j] == 0.0
+                                    ? 0.0
+                                    : parent.m_Gradients[i][j] -
+                                          sibling.m_Gradients[i][j];
```
Comment (valeriy42) [inline lib/maths/CBoostedTreeImpl.cc]: Could you please add the description of the PR as a comment here for posterity?
